File: memory/workspace.py
from __future__ import annotations

from pathlib import Path

_AGENTS_TEMPLATE = """\
# AGENTS.md

Operating instructions and persistent memory for this agent.

## Role
You are a helpful assistant with access to file, shell, and web tools.

## Memory
<!-- Add persistent notes here. This file is injected as your system prompt. -->
"""

_USER_TEMPLATE = """\
# USER.md

User profile and preferences.

## Preferences
<!-- Add user-specific preferences here. -->
"""


def bootstrap(workspace: Path) -> None:
    workspace.mkdir(parents=True, exist_ok=True)
    agents_md = workspace / "AGENTS.md"
    user_md = workspace / "USER.md"
    if not agents_md.exists():
        agents_md.write_text(_AGENTS_TEMPLATE)
    if not user_md.exists():
        user_md.write_text(_USER_TEMPLATE)


def build_system_prompt(workspace: Path, extra: str = "") -> str:
    parts = []
    for name in ("AGENTS.md", "USER.md"):
        path = workspace / name
        if path.exists():
            parts.append(path.read_text())
    if extra:
        parts.append(extra)
    return "\n\n---\n\n".join(parts)

File: src/memory/workspace.ts
import { mkdirSync, existsSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";

const AGENTS_TEMPLATE = `# AGENTS.md

Operating instructions and persistent memory for this agent.

## Role
You are a helpful assistant with access to file, shell, and web tools.

## Memory
<!-- Add persistent notes here. This file is injected as your system prompt. -->
`;

const USER_TEMPLATE = `# USER.md

User profile and preferences.

## Preferences
<!-- Add user-specific preferences here. -->
`;

export function bootstrap(workspace: string): void {
  mkdirSync(workspace, { recursive: true });
  const agentsMd = join(workspace, "AGENTS.md");
  const userMd = join(workspace, "USER.md");
  if (!existsSync(agentsMd)) {
    writeFileSync(agentsMd, AGENTS_TEMPLATE, "utf-8");
  }
  if (!existsSync(userMd)) {
    writeFileSync(userMd, USER_TEMPLATE, "utf-8");
  }
}

export function buildSystemPrompt(workspace: string, extra: string = ""): string {
  const parts: string[] = [];
  for (const name of ["AGENTS.md", "USER.md"]) {
    const path = join(workspace, name);
    if (existsSync(path)) {
      parts.push(readFileSync(path, "utf-8"));
    }
  }
  if (extra) parts.push(extra);
  return parts.join("\n\n---\n\n");
}
